src/devices/bluetooth.c

```diff
@@ -1,169 +1,230 @@
+}
+
+static int bt_set_name(DeviceConfig *config, const char *bt_name)
+{
+        pr_info_str_msg("BT: Setting name: ", bt_name);
+
+        char buf[BT_MAX_NAME_LEN + 7 + 1] = "AT+NAME";
+        strlcpy(buf + 7, bt_name, BT_MAX_NAME_LEN + 1);
+
+        return sendBtCommandWaitResponse(config, buf, "OKsetname",
+                                         COMMAND_WAIT);
+}
+
+static int bt_set_pin(DeviceConfig *config, const char *pin_str)
+{
+        pr_info_str_msg("BT: Setting pin: ", pin_str);
+
+        char buf[BT_MAX_PIN_LEN + 6 + 1] = "AT+PIN";
+        strlcpy(buf + 6, pin_str, BT_MAX_PIN_LEN + 1);
+
+        return sendBtCommandWaitResponse(config, buf, "OKsetPIN",
+                                         COMMAND_WAIT);
+}
+
+
+static bool bt_find_working_baud(DeviceConfig *config, const int targetBaud)
+{
+        pr_info("BT: Searching for working baud rate...\r\n");
+
+        const int rates[] = {targetBaud, 230400, 115200, 57600, 9600};
+        size_t i = 0;
+        for (; i < sizeof(rates); ++i) {
+                if(set_check_bt_serial_baud(config, rates[i]))
+                        break;
+        }
+
+        /* Check that we didn't fail to find a workable rate */
+        if (i == sizeof(rates)) {
+                pr_info("BT: Unable to communicate with device.\r\n");
+                return false;
+        }
+
+        pr_info_int_msg("BT: Talking with BT device.  Baud: ", rates[i]);
+        return true;
+}
+
+int bt_disconnect(DeviceConfig *config)
+{
+        return 0; //NOOP
+}
+
+int bt_init_connection(DeviceConfig *config)
+{
+        BluetoothConfig *btConfig =
+                &(getWorkingLoggerConfig()->ConnectivityConfigs.bluetoothConfig);
+        unsigned int targetBaud = btConfig->baudRate;
+        const char *deviceName = btConfig->deviceName;
+        const char *pin = btConfig->passcode;
+
+        /* give a chance for BT module to init */
+        delayMs(BT_INIT_DELAY);
+
+        /*
+         * The HC-O6 seems to sometimes have trouble dealing with long AT
+         * commands. Namely at high speed it seems that the device can't
+         * process certain AT commands fast enough, causing an empty response.
+         * To get past this, we set the baud rate while programming down to
+         * factory level (9600).  This ensures things go slow enough for the
+         * HC-06 processor + code to handle it.
+         */
+        const bool status =
+                bt_find_working_baud(config, targetBaud) &&
+                bt_set_baud(config, BT_AT_CMD_BAUD) &&
+                bt_get_version(config) &&
+                bt_set_name(config, deviceName) &&
+                bt_set_pin(config, pin) &&
+                bt_set_baud(config, targetBaud);
+
+        if (status) {
+                pr_info("BT: Init complete\r\n");
+                g_bluetooth_status = BT_STATUS_PROVISIONED;
+        } else {
+                pr_info("BT: Failed to provision module.\r\n");
```

Comment:
if app is already connected via BT it will not respond to AT commands and the provisioning will fail, falling to line 220. Line 158 provided a hint that this might be the case, saying we may already be connected. should update this line to be consistent with that.